third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/LoadStoreOpToLLVM.cpp

```diff
@@ -1454,78 +1458,177 @@ static LogicalResult iterateGatherScatterIndices(
   unsigned threadsPerWarp = xCoordsLayout.getInDimSize(kLane);
   unsigned numWarps = xCoordsLayout.getInDimSize(kWarp);
 
-  // Each gather4 instructions reads 128 bytes for 4 rows at a time.
+  // Suppose the result tensor is `tensor<MxNxdtype>`. Each gather4/scatter4
+  // message covers a `[4, contigDimSize]` tile of the result, where the `y`
+  // elements are contiguous but the `x` elements are separately indexed using
+  // indices load out such that 4 consecutive elements are grouped together.
+  //
+  // Because each warp can emit a distinct message, even if the layout of the
+  // indices are redundant along the warp bases, we can still figure out a way
+  // to issue different messages per instruction. E.g. consider the layout
+  //
+  //   register = [[1], [2]]
+  //   lane =     [[4], [8], [0], [0], [0]]
+  //   warp =     [[16], [0]]
+  //   block =    []
+  //   outDims: dim0 (size 32)
+  //
+  // If the tensor is `tensor<32x128xf32>`, there are a total of 32 tiles
+  // (4 rows per message, and in 1 CTA mode, 128B per row). If we mask out the
+  // redundant warps, we have to emit 16 instructions to fill the 32 tiles. We
+  // can "vectorize" the subtiles along the redundant warps to emit only 8
+  // instructions.
```

Comment:
Why don't we just pick a layout that doesn't duplicate data across warps? Doesn't that address the register usage more directly?

Reply:
You know, that might have been the simpler solution... now that I think about it. The complexity in the lowering is just compensating for this 🙃.

I guess the flipside is that having a more flexible lowering leaves the door open for picking layouts that are better for other reasons, such as load coalescing...

Reply:
I would strongly prefer to move the compicated layout finding logic out of the lowering itself. The operation you describe here would be a view-only layout conversion from whatever the current layout would be, so it should be equivalent to just add a convert layout to the appropriate linear layout in the ttgir.

Reply:
I don't think a convert_layout is equivalent because this logic also tiles in the Y direction. The only time that would be beneficial if we picked the right layout to begin with, however, is when BLOCK_M is tiny, like 8 or 16, but it should never be less than 64 in practice to use MMAv5. 

I'll go with your suggestion and delete the complicated work stealing logic from the lowering. I am still trying to measure the performance atm